### DISCUSSION TITLE
Ubuntu snap upgrade from zwavejs2mqtt to zwave-js-ui

### DISCUSSION THREAD

**Discussion by northrup** [zwave-js/zwave-js-ui#2787] [Q&A] [answered]
The latest versions published to ubuntu snaps are `7.2.0` as displayed here: 

```bash
snap-id:      k26XpW9UYHL1HNh20isi5MHoazqHZ8br
tracking:     latest/edge
refresh-date: yesterday at 21:14 UTC
channels:
  latest/stable:    v6.15.2       2022-08-30 (418) 142MB -
  latest/candidate: v7.1.0        2022-09-07 (434) 141MB -
  latest/beta:      v7.1.0-SEL-CC 2022-09-07 (438) 141MB -
  latest/edge:      v7.2.0        2022-10-12 (446) 146MB -
installed:          v7.2.0                   (446) 146MB -
root@ubuntu:~# 
```

Are there any plans to publish `8.2.1` to the official ubuntu snap repo?

**Comment by robertsLando** [comment]
@jmgiaever is working on it, I'm not maintaining the snap package

**Comment by northrup** [comment]
@jmgiaever anything I can do to help?

**Comment by jmgiaever** [comment]
https://github.com/zwave-js/zwave-js-ui/discussions/2670#discussioncomment-3835812

I haven't had time to solve the problems described. I just have had to much thing on my desk to test and debug. 😑

But I guess the issue is with the settings.json file, when it's copied, since there are references to the old `zwavejs2mqtt` paths in there. (If I remember correctly).

The other guy had to re-interview the whole network. 

If you're on Ubuntu, it s should be `/var/snap/zwave-js-ui` instead of `/var/snap/zwavejs2mqtt`. 

You can verify the path by loggin into the shell after installing zui. 

```
$ sudo snap run --shell zwave-js-ui
$ echo $SNAP_DATA
```
Then `exit` out from the shell. 

Please **disable the zwavejs2mqtt deamon** before proceeding (`zwavejs2mqtt.disable`)

Move all the content from (what I'm assuming is)
`/var/snap/zwavejs2mqtt/<revision>` to `/var/snap/zwave-js-ui/<revision>`

`<revision>` is the current running revision number your running. You can possibly also use `current` in the replacement for the number, but make sure it's mounted first. If not check current running revisions with `snap list | grep zwave`

E.g
```
ubuntu@home-assistant:~$ snap list | grep zwave
zwavejs2mqtt         v7.1.0-SEL-CC           438    latest/beta      giaever-online  -
```

My running revision is 438. You might have 2-3 revisions installed for rollback, but they might be outdated. 

Now edit the settings.json file in the new location (zwave-js-ui) with - what I assume is - `/var/snap/zwave-js-ui`. 

Please exclude copying over the `.ext-config`. Zwave-js will re-build this directory within seconds (no interview required). 

Please let me know if this solution still require zwave-js to do a full re-interview.

**Comment by jmgiaever** [comment]
Oh, I found some wrong paths. Suggest waiting upgrading.

**Comment by jmgiaever** [comment]
Hi, I think all the bugs should be fixed now; the thing missing is the copy command and some other things that is not necessary for the app to run. All of the missing stuff is for the «easy transformation» between the old and new.

**Comment by MichelHanzen** [comment]
I follow your advice and didn't have to re-interview the nodes. 
zwavejs part was working correctly after issuing the command "snap connect zwave-js-ui:raw-usb". 
To make the UI work, I needed to issue the commands:
snap set zwave-js-ui server.host=0.0.0.0
snap set zwave-js-ui server.ssl=true

**Comment by jmgiaever** [comment]
Hi,

Thank. I fixed most of the issues the other day, but it great to have it confirmed. 

I will soon release a stable release on it.

**Comment by chench00** [comment]
> I will soon release a stable release on it.
@jmgiaever If not too much to ask, could you please include some instructions on how to merge for novices like myself? I am terrified of breaking my network. Willing to help in any way i can if needed.

**Comment by jmgiaever** [comment]
Yes. I'm soon pushing the final files and you can do everything from the built-in commands, that I'm adding. 

It has to build and I have to do some testing. I'll get back to you.

**Comment by Whytey** [comment]
Hi, is this now OK to move from the zwavejs2mqtt snap to the zwave-js-ui snap?  

I have been without my Z-Wave devices since I upgraded to HA 2022.10 a week ago and have been hesitant to swap snaps since there is some discussion about having to re-interview networks, etc.

As an aside, whilst I saw the note about zwave-js-server in the [breaking changes](https://www.home-assistant.io/blog/2022/10/05/release-202210/#breaking-changes) - I struggled to correlate it with the version numbers of the server, etc provided by the snap - hopefully something we can address in the future?

Cheers for all you do!

**Comment by LordMike** [comment]
Just chiming in to say, @Whytey, that I did this move a few days ago, because ... "it's just files".. :P

* I installed the new `zwave-js-ui` snap (don't start it), and stopped the `zwavejs2mqtt` snap.
* I copied over my configurations via snap CLI (`snap get|set|list`), I didn't find an easy way so I just read each value out and `set` it on the new snap
* I copied over my files `/snap/zwavejs2mqtt/current/` to `/snap/zwave-js-ui/current/`
* I fixed some of the paths in the files, this comes from past experience, where there are some absolute paths that go to the proper revision of the snap. I ended up doing something like `grep -rF '/snap/zwavejs2mqtt/' /snap/zwave-js-ui/current/` to find these, and fix them individually
* There are some steps to allow certain features on the new snap, notably USB access
* Then I started the new snap - and it worked :)

**Comment by Whytey** [comment]
Thanks very much!!  I am now up and running and appear compatible with HA again :)

Output of the commands I executed - somewhat cleaned up to remove some of the rabbit holes I followed at random points.

```
root@zwave:~# snap install zwave-js-ui
zwave-js-ui v8.2.1 from Giaever.online (giaever-online) installed
root@zwave:~# snap disable zwavejs2mqtt
2022-10-25T21:15:46+10:00 INFO Waiting for "snap.zwavejs2mqtt.zwavejs2mqtt.service" to stop.
zwavejs2mqtt disabled
root@zwave:~# snap get zwavejs2mqtt
Key      Value
mqtt     {...}
server   {...}
session  {...}
root@zwave:~# snap get zwavejs2mqtt mqtt
Key        Value
mqtt.name  
root@zwave:~# snap get zwavejs2mqtt server
Key                       Value
server.force-disable-ssl  false
server.host               0.0.0.0
server.port               8091
server.ssl                false
root@zwave:~# snap set zwave-js-ui server.force-disable-ssl=false
root@zwave:~# snap set zwave-js-ui server.host=0.0.0.0
root@zwave:~# snap set zwave-js-ui server.port=8091
root@zwave:~# snap set zwave-js-ui server.ssl=false
root@zwave:~# snap get zwavejs2mqtt session
Key                    Value
session.cookie-secure  
session.secret         <redacted>
root@zwave:~# snap set zwave-js-ui session.secret=<redacted>
root@zwave:~# cd /var/snap/zwavejs2mqtt/433/
root@zwave:/var/snap/zwavejs2mqtt/433# ls
630b.jsonl                 zwavejs_2022-05-20.log.gz  zwavejs_2022-06-05.log.gz  zwavejs_2022-06-28.log.gz  zwavejs_2022-08-04.log.gz  zwavejs_2022-08-27.log.gz
630b.metadata.jsonl        zwavejs_2022-05-21.log.gz  zwavejs_2022-06-06.log.gz  zwavejs_2022-07-18.log.gz  zwavejs_2022-08-05.log.gz  zwavejs_2022-08-28.log.gz
630b.values.jsonl          zwavejs_2022-05-22.log.gz  zwavejs_2022-06-07.log.gz  zwavejs_2022-07-19.log.gz  zwavejs_2022-08-06.log.gz  zwavejs_2022-08-29.log.gz
config                     zwavejs_2022-05-23.log.gz  zwavejs_2022-06-08.log.gz  zwavejs_2022-07-20.log.gz  zwavejs_2022-08-07.log.gz  zwavejs_2022-08-30.log.gz
logs                       zwavejs_2022-05-24.log.gz  zwavejs_2022-06-09.log     zwavejs_2022-07-21.log.gz  zwavejs_2022-08-08.log.gz  zwavejs_2022-08-31.log
nodes.json                 zwavejs_2022-05-25.log.gz  zwavejs_2022-06-22.log.gz  zwavejs_2022-07-22.log.gz  zwavejs_2022-08-09.log.gz  zwavejs_current.log
sessions                   zwavejs_2022-05-26.log.gz  zwavejs_2022-06-23.log.gz  zwavejs_2022-07-23.log.gz  zwavejs_2022-08-10.log.gz  zwavejs_logrotate.json
settings.json              zwavejs_2022-05-27.log.gz  zwavejs_2022-06-24.log.gz  zwavejs_2022-07-24.log.gz  zwavejs_2022-08-11.log
snippets                   zwavejs_2022-06-02.log.gz  zwavejs_2022-06-25.log.gz  zwavejs_2022-08-01.log.gz  zwavejs_2022-08-24.log.gz
users.json                 zwavejs_2022-06-03.log.gz  zwavejs_2022-06-26.log.gz  zwavejs_2022-08-02.log.gz  zwavejs_2022-08-25.log.gz
zwavejs_2022-05-19.log.gz  zwavejs_2022-06-04.log.gz  zwavejs_2022-06-27.log.gz  zwavejs_2022-08-03.log.gz  zwavejs_2022-08-26.log.gz
root@zwave:/var/snap/zwavejs2mqtt/433# cp -r * /var/snap/zwave-js-ui/34/
root@zwave:/var/snap/zwave-js-ui/34# grep -rF '/snap/zwavejs2mqtt/' .
./zwavejs_2022-08-11.log:2022-08-11T10:47:44.159Z CONFIG   Using external configuration dir /var/snap/zwavejs2mqtt/414/.ext-config
./zwavejs_logrotate.json:    "auditLog": "/var/snap/zwavejs2mqtt/393/zwavejs_logrotate.json",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-01.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-02.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-03.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-04.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-05.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-06.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-07.log",
./zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/393/zwavejs_2022-08-08.log",
./logs/zwavejs_logrotate.json:    "auditLog": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_logrotate.json",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-02.log",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-03.log",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-04.log",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-05.log",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-06.log",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-07.log",
./logs/zwavejs_logrotate.json:            "name": "/var/snap/zwavejs2mqtt/419/logs/zwavejs_2022-09-08.log",
./settings.json:    "deviceConfigPriorityDir": "/var/snap/zwavejs2mqtt/433/config",
root@zwave:/var/snap/zwave-js-ui/34# nano settings.json 
<<Edited the file to swap the path to zwave-js-ui>>
<<Did not edit the logrotate.json files, hope that doesn't come to bite me later>>
root@zwave:/var/snap/zwave-js-ui/34# snap connect zwave-js-ui:raw-usb
root@zwave:/var/snap/zwave-js-ui/34# snap connect zwave-js-ui:hardware-observe
root@zwave:/var/snap/zwave-js-ui/34# snap start zwave-js-ui
Started.
root@zwave:/var/snap/zwave-js-ui/34# snap logs zwave-js-ui
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]: 2022-10-25 21:41:41.540 INFO APP: Application path:/snap/zwave-js-ui/34/lib/zwave-js-ui
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]:   ______  __          __                      _  _____     _    _ _____
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]:  |___  /  \ \        / /                     | |/ ____|   | |  | |_   _|
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]:     / /____\ \  /\  / /_ ___   _____         | | (___     | |  | | | |
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]:    / /______\ \/  \/ / _' \ \ / / _ \    _   | |\___ \    | |  | | | |
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]:   / /__      \  /\  / (_| |\ V /  __/   | |__| |____) |   | |__| |_| |_
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]:  /_____|      \/  \/ \__,_| \_/ \___|    \____/|_____/     \____/|_____|
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]: 2022-10-25 21:41:41.564 WARN STORE: scenes.json not found
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]: Logging to file:
2022-10-25T21:41:41+10:00 zwave-js-ui.zwave-js-ui[957855]: 	/var/snap/zwave-js-ui/34/logs/zwavejs/zwavejs_2022-10-25.log
root@zwave:/var/snap/zwave-js-ui/34# 
```

**Comment by chench00** [comment]
While I appreciate all of these helpful comments, I would still prefer a solution from @jmgiaever (no hurry of course) as I am terrified of having to redo everything if I fail. ;-)

**Comment by jmgiaever** [comment]
I'm still waiting for some stuff from the Snap team. 

But you should be fine following this, as long as you copy from z2m. If something fails in ZUI you can just disable ZUI and enable Z2M again.

**Comment by chench00** [comment]
@Whytey I went ahead and followed your detailed instructions (thank you by the way) and have two questions:

    root@zwave:/var/snap/zwave-js-ui/34# nano settings.json 
    Edited the file to swap the path to zwave-js-ui

I this case, the path is `/var/snap/zwave-js-ui/34/config` ?

        Did not edit the logrotate.json files, hope that doesn't come to bite me later

Is this something I should do/worry about? I am willing to change mine but not sure exactly which line to change. Here's what mine looks like:

        {
            "keep": {
                "days": true,
                "amount": 7
            },
            "auditLog": "/var/snap/zwavejs2mqtt/391/zwavejs_logrotate.json",
            "files": [
                {
                    "date": 1656734400004,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-02.log",
                    "hash": "83886311f75bc20a415bc75ab1c2451dd7dce1b2910fd9bd5b69312c6f20eaae"
                },
                {
                    "date": 1656820800013,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-03.log",
                    "hash": "29e20209da00f7fdebc72fad79ec85a9d89eeaf339b4f88036c73fc96ef61866"
                },
                {
                    "date": 1656907200008,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-04.log",
                    "hash": "88ff8261f85e36a8bd94b85d284c951383483ea7aec7fd57e232398f1bae4117"
                },
                {
                    "date": 1656993600007,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-05.log",
                    "hash": "f170d396c3a824a076e98fcb3aacb50094a72c1abfa2024b78e43593cd7f839d"
                },
                {
                    "date": 1657080000039,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-06.log",
                    "hash": "cf95bcd51b3376c5c61051a5d379b24387b5d859c2c2f2308dad832e1831a757"
                },
                {
                    "date": 1657166400026,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-07.log",
                    "hash": "27d9fc8399ac9853a57aef248632ef2fcd7955d045df6e37443daf2325e9ec4b"
                },
                {
                    "date": 1657252800015,
                    "name": "/var/snap/zwavejs2mqtt/391/zwavejs_2022-07-08.log",
                    "hash": "5195790902c93a157fbf242822cd8ecd6cb443f0a9f783fa6308dd848e6e2f8a"
                }
            ],
            "hashType": "sha256"

**Comment by Whytey** [comment]
Hi @chench00 ,

That updated path looks correct, yes.  I don't think you need to worry about the logrotate changes, seems to be working fine for me.

Good luck,
Whytey

**Comment by pascalspits** [comment]
I also succeeded in upgrading to ZUI, following these instructions ... thanks a million !
should I now remove the old Z2M Snap ?
what will happen on a Raspi reboot, will the old Z2M or the new ZUI start up ?

I'm confused ...

**Comment by jmgiaever** [comment, marked as answer]
Just disable z2m and enable zui:

```
sudo zwavejs2mqtt.disable
sudo zwave-js-ui.enable
```

If everything runs ok in zui, I see no reason to keep z2m:

```
sudo snap remove --purge zwavejs2mqtt
```

Remember that release cycles will be the same for the channel/tracks for zui, so those who wants rolling releases and not just final stable releases should be on the candidate track. 

```
sudo snap switch zwave-js-ui --channel=candidate
```

**Comment by pascalspits** [comment]
how do I set ZUI "autostart" on boot ?

**Reply by robertsLando** [reply]
https://askubuntu.com/questions/1212260/how-do-i-keep-a-snap-from-starting-on-boot

**Reply by jmgiaever** [reply]
Or just the built-in command `sudo zwave-js-ui.enable` and you're good. ;)

**Reply by pascalspits** [reply]
So, if I once use this "enable" command, it will always start on boot ?

**Reply by pascalspits** [reply]
after every automatic update (like yesterday) I need to restart the service with SSH "sudo snap start zwave-js-ui" on th Raspi ... am I doing something wrong ?

**Reply by jmgiaever** [reply]
Why do you feel the need to restart, when the application just restarted itself?

**Reply by Whytey** [reply]
I found that even calling `snap enable zwave-js-ui` didn't actually enable the auto restart as I expected. 

After having noticed the same problem of the snap not restarting after a refresh, I eventually found the `zwave-js-ui.enable` command. 

I am hoping this has fixed things for me  👍

**Reply by jmgiaever** [reply]
If `snap enable` didn't work, then you have an issue with your setup. What the `.enable` command actually do is basically issuing the`snap enable` command.

**Reply by pascalspits** [reply]
exactly, it doesn't seem to restart itself after auto-upgrading ... so I wonder what I'm doing wrong ...

**Reply by pascalspits** [reply]
could this be my problem ?

```
pi@raspberrypi:~ $ sudo zwave-js-ui.enable
sudo: zwave-js-ui.enable: command not found
```

**Reply by pascalspits** [reply]
```
pi@raspberrypi:~ $ sudo snap enable zwave-js-ui
error: cannot enable "zwave-js-ui": snap "zwave-js-ui" already enabled
```

**Reply by pascalspits** [reply]
today I had a power failure, and the Raspi rebooted when power came back on ... and ZUI again didn't start ...
I had to SSH into the Raspi to start ZUI ...
so it's not only after upgrading, it's also on booting it doesn't autostart ...
the "sudo zwave-js-ui.enable" command doesn't seem to work on my system ...
what am I doing wrong ?

**Reply by jmgiaever** [reply]
Which OS is this? I think you have something messed up with the Snapd installation, and not with this particular snap.

**Reply by pascalspits** [reply]
Linux raspberrypi 5.15.32-v8+ #1538 SMP PREEMPT Thu Mar 31 19:40:39 BST 2022 aarch64

**Reply by jmgiaever** [reply]
That means Rasbian? If so, see: [the hack for Rasbian/Debian](https://github.com/zwave-js/zwave-js-ui/discussions/1216#discussioncomment-759534)

**Reply by pascalspits** [reply]
pi@raspberrypi:~ $ sudo snap run --shell zwavejs2mqtt
error: cannot find current revision for snap zwavejs2mqtt: readlink /snap/zwavejs2mqtt/current: no such file or directory

... because Z2M is already uninstalled ?

"/var/lib/snapd/hostfs/" is empty ... so line 2 also doesn't work ...

the secure path has been added now, but still not autostarting after boot ...

pi@raspberrypi:~ $ sudo snap start zwave-js-ui
Started.

manually starting it works fine ... weird ...

**Reply by jmgiaever** [reply]
You'll naturally have to swap out every reference to «z2m» with «zui».
